### ISSUE TITLE
Support `parquet` columnar format in the `aws_s3` sink

### ISSUE THREAD

**Issue by binarylogic** [timberio/vector#1374]
Similar to https://github.com/timberio/vector/issues/1373 we should support the [Parquet format](https://parquet.apache.org/).  The `parquet` format is a columnar format that enables faster and more efficient data access schemes such as column selection and indexing.

## Implementation

Unfortunately, I do not have deep experience with this format as I do with ORC, but like everything else, we should start very simple. Fortunately, there appears to be a [Rust library](https://github.com/sunchao/parquet-rs) that supports basic writing of this data.

**Comment by LucioFranco** [comment]
One issue with that parquet library is that it requires nightly rust https://github.com/sunchao/parquet-rs/blob/master/src/lib.rs#L123 and https://github.com/apache/arrow/blob/master/rust/parquet/src/lib.rs#L18, we can't use these features because we are on the stable compiler.

**Comment by binarylogic** [comment]
Any idea what the timeline is for those to be on stable?

**Comment by LucioFranco** [comment]
Well really `specialization` is the last one on the updated `apache/arrow` but that feature has been in progress since 2016, so we will likely not see it any time soon. Our best bet would be to attempt to remove that feature from that library but I am not sure how hard that would be or if that is even possible.

**Comment by camerondavison** [comment]
I think that it still requires nightly, but just throwing into the mix https://github.com/apache/arrow/tree/master/rust/parquet

**Comment by davemt** [comment]
Progress in Apache Arrow project:

- PR: Remove rust specialization - https://github.com/apache/arrow/pull/8698
- JIRA: Support stable Rust - https://issues.apache.org/jira/browse/ARROW-6717
- JIRA (comment): re. remaining dependency other than specialization - https://issues.apache.org/jira/browse/ARROW-6718?focusedCommentId=17002842&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17002842

**Comment by Alexx-G** [comment]
IIRC Apache Arrow now works on stable Rust 🎉 (unless SIMD feature is used). Would it be possible to re-evaluate this one?
In case the team doesn't have bandwidth, but somebody can help with review and some pointers, I can look into it.

**Comment by jszwedko** [comment]
Hi @Alexx-G . This hasn't been scheduled yet, but you wanted to pick it up, I'd be happy to advise/review!

**Comment by Alexx-G** [comment]
I've checked the `parquet` crate it _seems_ to have some basic support for writing data.
@binarylogic @jszwedko I see that the ORC issue has some additional context, do you have any insights/pointers for this one? Especially about schema and the changes around the current encoding logic.
I'll try to run some tests with Athena+S3 and `parquet` to see whether I can get an MVP.

**Comment by corbyshaner** [comment]
Is there an update on this?  I'm very interested in JSON input to parquet output to S3.

**Comment by bhataprameya** [comment]
Any update on this? I am also interested in  JSON input and Parquet output on s3

**Comment by jgournet** [comment]
Any chance this can be looked at ? 
Right now, we send logs from vector to Kinesis just to get them converted to Parquet... I'd be happy to skip that costly step

**Comment by spencergilbert** [comment]
This codec isn't currently on our roadmap, but we could help a community PR get merged 👍

**Comment by ktff** [comment]
Hey @fuchsnj 

Regarding 1. question, no, a Parquet batch can't be built from already encoded events. It's necessary to intercept them before that, or process them in a suitable way for Parquet. Fortunately there is a way to to that, by implementing https://github.com/vectordotdev/vector/blob/018637ad93ebb74bcec597e89d33127ac83202d8/src/sinks/util/encoding.rs#L9 for `Vec<Event>`. 

There are similar cases:
* https://github.com/vectordotdev/vector/blob/018637ad93ebb74bcec597e89d33127ac83202d8/src/sinks/elasticsearch/encoder.rs#L56
* https://github.com/vectordotdev/vector/blob/018637ad93ebb74bcec597e89d33127ac83202d8/src/sinks/splunk_hec/logs/encoder.rs#L61
* https://github.com/vectordotdev/vector/blob/018637ad93ebb74bcec597e89d33127ac83202d8/src/sinks/gcp/chronicle_unstructured.rs#L297
* https://github.com/vectordotdev/vector/blob/018637ad93ebb74bcec597e89d33127ac83202d8/src/sinks/datadog/logs/sink.rs#L127

Current `aws_s3` sink is using https://github.com/vectordotdev/vector/blob/018637ad93ebb74bcec597e89d33127ac83202d8/src/sinks/util/encoding.rs#L18 which would be replaced by `ParquetEncoder` when it's configured. 

---

Regarding 2. question, Parquet is a file format that is usually the final/long term destination for storing data that can later be read for queries, by ,for example, Amazon Athena. Such systems will/do encounter differences in schema with time, especially if they are performing federated queries, so they usually have the means to reconcile different schemas. That said, I came to this conclusion via research so comments from those with experience would be much appreciated.

So my main argument is that it's better to have the event reach it's destination and leave it to the user to transform the event into desired schema before or after the sink if they so wish/need to, then to require configuration of fixed schema that drops events which forces those that do have events with varying schema to yet again define the schema in some transformer before this sink to transform the events or to change the fixed schema.

**Comment by fuchsnj** [comment]
1. The `impl Encoder<Vec<ProcessedEvent>>` is the sink specific encoder. While you _could_ have a setting that replaces this on the sink, this doesn't follow our current conventions, and would only work for that specific sink. What we would prefer (which isn't quite supported yet) is expanding our codecs to add `parquet`, which would allow this to be used in any sink. The current (single event only) codecs are selected [here](https://github.com/vectordotdev/vector/blob/018637ad93ebb74bcec597e89d33127ac83202d8/src/sinks/aws_s3/config.rs#L104). 

2. >So my main argument is that it's better to have the event reach it's destination and leave it to the user to transform the event into desired schema before or after the sink if they so wish/need to, then to require configuration of fixed schema that drops events which forces those that do have events with varying schema to yet again define the schema in some transformer before this sink to transform the events or to change the fixed schema.

   Our current codecs define a `schema_requirement` which can type check events at Vector startup to ensure events match the    expected parquet schema. This would prevent events from being dropped at runtime for not matching the schema. (Note that full schema support in Vector is relatively new, and not enabled by default yet (It's the global setting `schema.enabled`)).

**Comment by jszwedko** [comment]
For point 1, I'm in agreement with @fuchsnj . It'd be _really_ nice if we could update the codec model to do the batch encoding in a generic way to work across all sinks using codecs. This came up recently with the `avro` codec not outputting the schema at the head of each batch (https://github.com/vectordotdev/vector/issues/16994) and would come up with other codecs that require headers like `csv`.

**Comment by fuchsnj** [comment]
The work for adding support for "batched" codecs is discussed here: https://github.com/vectordotdev/vector/issues/16637

**Comment by ktff** [comment]
For point 1., while implementation of `parquet` would be sink specific encoder at the moment, it can be written to be generic in a sense that it can be reused in a different sink and/or in the batch encoder abstraction/feature once that comes. 

Going by https://github.com/vectordotdev/vector/issues/16637 for `parquet` codec envelope is not enough, it's column vise format so it needs to deconstruct all events to form those columns. That is inline with https://github.com/vectordotdev/vector/issues/16637#issuecomment-1462691621 for batch sinks.

If batch encoding was already implemented that would simplify things. Going by `impl` diversity of `RequestBuilder` that's not something trivially doable. From what I see there are two parts to it. Abstraction over encoders and abstraction over events. For `parquet` codec in `aws_s3` sink, abstraction over encoders is necessary. Just using `Box<dyn Encoder<E>>` seems fine for that. The second abstraction is the one that would allow for this codec to be used in other sinks. I would say that that is an orthogonal concern/feature since it would need to unify all types of events in batch sinks into a single thing and then modify, if necessary, all of the codecs to use this unified type. That's a completely other issue.

For point 2. @fuchsnj  can you provide some docs for `schema_requirement`, I can't find it in the docs. But that feature would be useful.

**Comment by davemt** [comment]
We currently collect logs and convert and store as parquet-on-s3, using fluentd and an internal plugin we manage for the parquet conversion on our aggregators. We are very interested in migrating to Vector and this issue is currently the remaining blocker.

FWIW, in our setup we configure a schema to use for each type of log and we would hope that the Vector implementation would at least support an option for specifying schemas.

**Comment by ktff** [comment]
@fuchsnj found the `schema_requirement`. That seems like exactly what's needed. So we can go with that.

Instead of determining schema during runtime, add option to specify schema for passing events.

**Comment by ktff** [comment]
@spencergilbert I do. I was on vacation hence the silence. The 1. point raised by @fuchsnj remains uresolved. Simplified, there are two ways forward:
1. Implement `parquet` codec only for `aws_s3` sink. Once the support for batched codecs lands, that limitation can be lifted.
2. Wait for batched codecs support and then merge `parquet` codec. 

In both cases, my plan is to implement codec first and in the case of 2. submit it once batched codecs have landed.

**Comment by spencergilbert** [comment]
Hope you had a nice vacation! Sounds good to me.

**Comment by Kikkon** [comment]
@ktff  I happened to find a similar solution. If you need help anywhere, I'd be happy to contribute together! If I can figure it out, that is. 🚀

**Comment by Kikkon** [comment]
> @Kikkon the draft contains functioning `parquet` codec for `aws_s3` sink. It's missing two important things though:
> 
> * Proper dealing with non conforming events. This can be resolved with accepting a performance hit.
> * Support for batched codecs. This is the only blocking issue.
> 
> I'm currently not in the position to add support for batched codecs hence the limbo state of #17395. If this is something you feel confident to add to Vector then reach out to @jszwedko. Once that lands I'll be able to finish the PR and get it merged.

I have some experience using Parquet, but am not familiar with Vector. The issue with this PR is: Parquet does not support append writes. If appending, it may require merging new and old files which has performance costs. However, batch implementation does not yet exist in Vector. So for now the PR is pending. if right?

@jszwedko  The Vector community has plans for a proposal to support batch writes. Perhaps once I become more familiar with Vector's architecture, I can discuss with everyone how we could add batch write.

**Comment by jszwedko** [comment]
> @jszwedko The Vector community has plans for a proposal to support batch codecs. Perhaps once I become more familiar with Vector's architecture, I can discuss with everyone how we could add batch codecs.

Hey! Yes, we would like to add the concept of batches to codecs but haven't been able to prioritize it on our side just yet. We'd be happy to help guide a contribution for it. I believe @lukesteensen would have some thoughts about what it could look like and also be able to answer questions.

**Comment by Kikkon** [comment]
Can you give me some advice? @lukesteensen  🫡

**Comment by ktff** [comment]
>The issue with this PR is: Parquet does not support append writes. If appending, it may require merging new and old files which has performance costs. However, batch codecs implementation does not yet exist in Vector. So for now the PR is pending. if right?

@Kikkon yes. Also one thing to note, currently the PR does add limited support for batch codecs to support Parquet but it's hacky and so not something that can be merged. The goal is to replace this with proper support.

**Comment by Kikkon** [comment]
> > The issue with this PR is: Parquet does not support append writes. If appending, it may require merging new and old files which has performance costs. However, batch codecs implementation does not yet exist in Vector. So for now the PR is pending. if right?
> 
> @Kikkon it's not doable with regular codecs, but if it would be then yes it would have performance cost. Also one thing to note, currently the PR does add limited support for batch codecs to support Parquet but it's hacky and so not something that can be merged. The goal is to replace this with proper support.

Got it, thank you！

**Comment by lukesteensen** [comment]
@Kikkon that's correct, you can't really do this right now with the current design of our codecs crate. It's something I would love to enable, but I haven't had time to figure out a good path forward. It will likely take quite a bit of refactoring and design work. I'm hoping we'll be able to tackle that before long, as I agree this is a feature we should have!

**Comment by Kikkon** [comment]
> 

@lukesteensen  Is there a corresponding roadmap currently? Are there any parts that I can participate in?

**Comment by rstml** [comment]
This is a really big blocker for us. Any plans to make this happen?

**Comment by Ralkion** [comment]
@rstml : We've used the patch provided by @ktff and applied it ourselves to a release branch (can't remember which one off the top of my head). It's been working fine for us for a while now. It only works for S3, but it does work.

**Comment by rstml** [comment]
@Ralkion thanks! that's helpful. Will give it a try.

I got an impression that #16637 was a blocker, but it seems that we should be able to define metadata during consumption.

Why not to merge #17395 then? If there's a support for Avro that works without headers, Parquet should be acceptable too?

**Comment by Ralkion** [comment]
My understanding from the discussion above (and others may correct me if I'm over-simplifying) is that @ktff 's patch only works for S3, and only works in a certain way that isn't necessarily applicable to other output destinations. Because it can't be generalized to other destinations it's not really 'production ready'.

**Comment by bruceg** [comment]
@rstml We have not been able to prioritize it codec batches at this point, nor is it likely in the near future, unfortunately.

**Comment by ktff** [comment]
@rstml it's as @Ralkion said, this codec is currently accessible only through `aws_s3` sink. For it to be a general codec for all sinks codec batches need to be implemented but they aren't on the roadmap for the near future as @bruceg mentioned.

**Comment by fpytloun** [comment]
I am also very interested in this feature. Going to be on roadmap anytime soon?

**Comment by yjagdale** [comment]
@jszwedko  - any possibility to get this in roadmap?

**Comment by bharathiram** [comment]
Any update on this feature?

**Comment by pront** [comment]
We are aware there is a lot of demand for this feature but it's not something we currently have in our roadmap. We would be happy to review a community contribution.

**Comment by kbespalov** [comment]
Any updates?

**Comment by pront** [comment]
For those interested in this feature, @benjamin-awd's PR is significant progress. https://github.com/vectordotdev/vector/pull/24124#issuecomment-3474072047

**Comment by srstrickland** [comment]
Hey all- happy to see progress here.  I'm curious what the appetite for automatic schema inference might be for this feature?  I see it was discussed a while back near the top of the thread.

A static schema is reasonable (and desirable) for cases where we can coerce into a common format (e.g. the OTel data model), or it arrives in a consistent format (e.g. a highly customized pipeline).  But the vector data model is flexible, and we've built relatively large scale, multi-tenant pipelines to deliver telemetry of varying schemas.  Having to coerce the various schemas can be difficult/impossible without sacrificing either:
* Input flexibility: keeping pipeline schemas up to date is only feasible when there are a few unique schemas, and starts to fall down when trying to maintain a multi-tenant pipeline where schemas may evolve quickly or
* Parquet / columnar data optimal layouts: data compresses far better when schemas are specific to their logical dataset; coercing to something like the OTel data model (where all attributes are packed into a generic map where the values often have their own string-based encoding) strips away so much of the columnar benefits.

Our use case might be a bit esoteric: we have a flexible data pipeline built from vector components, with eventual destinations in Clickhouse and a few data lake formats (iceberg, delta).  The pipeline itself supports hundreds of logical datasets, each of which requires its own schema management / evolution.  Today, we rely on an external component which does this schema evolution (with our own custom constraints and patterns) which sits between vector & the final storage, creating tables & adding columns as needed based on the data coming through for a particular dataset.  I don't envision a world where that component disappears entirely, because _how_ tables get created or columns added is getting outside the bounds of what vector should be responsible for (I raised this before in https://github.com/vectordotdev/vector/issues/19993 and that was the agreed conclusion).  However, we currently have to rely on JSON communication between vector and the schema evolution component, and I think there is significant upside if vector can perform this inference rather than how we have to do it today.  the data over the wire will be far less, the schema manager won't have to scan all the data, and the data will require very little (if any) transformation for the final hop.

I'd be happy to take a first stab at this, but wanted to see first if it would be something folks are interested in or would support such a contribution.  Some thoughts I have around this:

* Schema inference should be optional & disabled by default - maintain current functionality to error out if not enabled and no schema supplied
* If schema inference is enabled, we should likely write all columns as nullable unless explicitly identified as non-nullable. We could accept a list of fields that _must_ be present and therefore should be in the schema as non-nullable.  Perhaps we can reuse the `schema` configuration as it exists from this PR to indicate those fields, and the inference only detects other columns that exist in the batch.
* Schema inference can _still_ be a foot-gun and it's still up to pipeline owners to transform the data in a way that won't cause collisions (e.g. something similar to VRL's `tag_external_types`) - you don't want to end up with parquet files that disagree on what "fieldX" is.  However, this danger is still present with manual schema specification if not careful.
* We likely need to do inference for every batch.
* There are probably several methods we could use to do the inference (e.g. "fast": use the first type seen for each unique fieldname; "safe": scan all the types seen for each fieldname and make a reasonable decision, etc.)
* How to handle complex types?  Structs vs maps?  Lists?  Lists of different types?  Etc.

Happy to pull this out into a separate issue as already my comment is larger than I expected it to be when I started writing it 😄  ... just wanted to see what people think of all this.

**Comment by szibis** [comment]
I have also implemented, based on this discussion, Parquet in S3 in this PR https://github.com/vectordotdev/vector/pull/24706

**Comment by pront** [comment]
Implemented by: https://github.com/vectordotdev/vector/pull/25156
Not exposed yet due to a bug: https://github.com/vectordotdev/vector/pull/25321
But will be available in the next nightly and in the upcoming release.
